pkg/driver/node_server.go

```diff
@@ -1654,7 +1656,62 @@ func (driver *Driver) NodeGetVolumeStats(ctx context.Context, in *csi.NodeGetVol
 	log.Trace(">>>>> NodeGetVolumeStats")
 	defer log.Trace("<<<<< NodeGetVolumeStats")
 
-	return nil, status.Error(codes.Unimplemented, "")
+	volumePath := in.GetVolumePath()
+	if volumePath == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "volume path %s is empty", volumePath)
+	}
+
+	// check if it is a mount point
+	dummy := mount.New("")
+	notMount, err := dummy.IsLikelyNotMountPoint(volumePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, status.Errorf(codes.InvalidArgument, "volume path %s does not exist", volumePath)
+		}
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if notMount {
+		return nil, status.Errorf(codes.InvalidArgument, "volume path %s is not mounted", volumePath)
+	}
+
+	folders := strings.Split(in.VolumePath, "/")
+	pvName := folders[len(folders)-2]
+	credentials, err := driver.flavor.GetCredentialsFromVolume(pvName)
+	if err != nil {
+		log.Error("err: ", err.Error())
+		return nil, status.Errorf(codes.Unavailable, "Failed to get credentials from volume with name %s, err: %s",
+			pvName, err.Error())
+	}
+
+	// Get storage provider using secrets
+	storageProvider, err := driver.GetStorageProvider(credentials)
+	if err != nil {
+		log.Error("err: ", err.Error())
+		return nil, status.Errorf(codes.Unavailable, "Failed to get storage provider from secrets for volume with id %s, err: %s",
+			in.VolumeId, err.Error())
+	}
+	// Fetch the volume using volume ID
+	volume, err := storageProvider.GetVolume(in.VolumeId)
+	if err != nil {
+		log.Error("err: ", err.Error())
+		return nil, status.Errorf(codes.Internal, "Error while retrieving volume with id %s from the backend, err: %s",
+			in.VolumeId, err.Error())
+	}
+	if volume == nil {
+		log.Infof("Volume with ID %s not found on the backend, return ",
+			in.VolumeId)
+		return nil, nil
+	}
+	return &csi.NodeGetVolumeStatsResponse{
+		Usage: []*csi.VolumeUsage{
+			{
+				Available: volume.FreeBytes,
+				Total:     volume.Size,
+				Used:      volume.UsedBytes,
+				Unit:      csi.VolumeUsage_BYTES,
+			},
+		},
+	}, nil
```

Comment:
Please add inode usage as well.

Reply:
Currently  I am getting these fields from csp
 "id" : "064763f97c7c986c6600000000000000000000000f",
  "name" : "pvc-ae6b8013-61cb-4794-844a-468eb89b5ca6",
  "size" : 10737418240,
  "description" : "Volume from csi",
  "published" : true,
  "config" : {
    "pool" : "default",
    "folder" : "",
    "encrypted" : false,
    "thick" : false,
    "performance_policy" : "SQL Server",
    "dedupe_enabled" : false,
    "limit_iops" : 76800,
    "limit_mbps" : -1,
    "destroy_on_delete" : false,
    "sync_on_detach" : false,
    "clone_of" : "",
    "create_snapshot" : false,
    "target_scope" : "volume"
  },
  "snaphotUsedBytes" : 106496,
  "pool_name" : "default",
  "base_snapshot_id" : "",
  "volume_group_id" : "",
  "used_bytes" : 0,
  "free_bytes" : 10737418240,
  "reservation_ratio" : 0.0,
  "data_compression_saved_bytes" : 18567860,
  "snapshot_used_bytes" : 106496
}

Do I need to get inode get from array or is it one of the values already pulled?

Reply:
Those don't come from the array.  See here: https://github.com/hpe-storage/csi-driver/pull/190/files

Reply:
Made the changes.